File: src/org/nextras/orm/intellij/reference/EntityPropertyReference.java
package org.nextras.orm.intellij.reference;

import com.intellij.psi.PsiElement;
import com.intellij.psi.PsiPolyVariantReferenceBase;
import com.intellij.psi.PsiReference;
import com.intellij.psi.ResolveResult;
import com.jetbrains.php.PhpIndex;
import com.jetbrains.php.PhpIndexImpl;
import com.jetbrains.php.lang.documentation.phpdoc.psi.PhpDocProperty;
import com.jetbrains.php.lang.psi.elements.Field;
import com.jetbrains.php.lang.psi.elements.MethodReference;
import com.jetbrains.php.lang.psi.elements.PhpClass;
import com.jetbrains.php.lang.psi.elements.StringLiteralExpression;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;
import org.nextras.orm.intellij.utils.OrmUtils;
import org.nextras.orm.intellij.utils.PhpClassUtils;
import org.nextras.orm.intellij.utils.PhpIndexUtils;

import java.util.ArrayList;

public class EntityPropertyReference extends PsiPolyVariantReferenceBase<StringLiteralExpression>
{

	public EntityPropertyReference(StringLiteralExpression psiElement)
	{
		super(psiElement);
	}

	@NotNull
	@Override
	public ResolveResult[] multiResolve(boolean b)
	{
		StringLiteralExpression expr = getElement();
		assert (expr.getParent().getParent() != null);

		MethodReference method = (MethodReference) expr.getParent().getParent();
		assert (method.getName() != null && (method.getName().equals("setValue") || method.getName().equals("setReadOnlyValue")));

		ArrayList<ResolveResult> result = new ArrayList<ResolveResult>();
		PhpIndex phpIndex = PhpIndex.getInstance(this.getElement().getProject());
		for (PhpClass cls : PhpIndexUtils.getByType(method.getClassReference().getType(), phpIndex)) {
			if (!OrmUtils.isEntity(cls, phpIndex)) {
				continue;
			}
			final Field field = cls.findFieldByName(expr.getContents(), false);
			if (field == null || !(field instanceof PhpDocProperty)) {
				continue;
			}
			result.add(new ResolveResult()
			{
				@Nullable
				@Override
				public PsiElement getElement()
				{
					return field;
				}

				@Override
				public boolean isValidResult()
				{
					return true;
				}
			});
		}
		return result.toArray(new ResolveResult[result.size()]);
	}

	@NotNull
	@Override
	public Object[] getVariants()
	{
		return new Object[0];
	}
}

File: src/org/nextras/orm/intellij/reference/EntityPropertyReference.kt
package org.nextras.orm.intellij.reference

import com.intellij.psi.PsiElement
import com.intellij.psi.PsiPolyVariantReferenceBase
import com.intellij.psi.ResolveResult
import com.jetbrains.php.PhpIndex
import com.jetbrains.php.lang.documentation.phpdoc.psi.PhpDocProperty
import com.jetbrains.php.lang.psi.elements.MethodReference
import com.jetbrains.php.lang.psi.elements.StringLiteralExpression
import org.nextras.orm.intellij.utils.OrmUtils
import org.nextras.orm.intellij.utils.PhpIndexUtils
import java.util.*

class EntityPropertyReference(psiElement: StringLiteralExpression) : PsiPolyVariantReferenceBase<StringLiteralExpression>(psiElement) {
	override fun multiResolve(b: Boolean): Array<ResolveResult> {
		val expr = element
		assert(expr.parent.parent != null)

		val method = expr.parent.parent as MethodReference
		assert(method.name != null && (method.name == "setValue" || method.name == "setReadOnlyValue"))

		val result = ArrayList<ResolveResult>()
		val phpIndex = PhpIndex.getInstance(this.element.project)
		for (cls in PhpIndexUtils.getByType(method.classReference!!.type, phpIndex)) {
			if (!OrmUtils.OrmClass.ENTITY.`is`(cls, phpIndex)) {
				continue
			}
			val field = cls.findFieldByName(expr.contents, false)
			if (field == null || field !is PhpDocProperty) {
				continue
			}
			result.add(object : ResolveResult {
				override fun getElement(): PsiElement? {
					return field
				}

				override fun isValidResult(): Boolean {
					return true
				}
			})
		}
		return result.toTypedArray()
	}

	override fun getVariants(): Array<Any> {
		return emptyArray()
	}
}
